Add support for hiding the Toolbar on Android

MattDavies opened the pull request:
This PR will ideally also make it a lot easier to support multiple Toolbar styles within a single app.

Most of the edge cases I could find were around multiple modals, but worth checking tab/single screen activity push/pop shows/hides the Toolbar correctly as well.

Haven't touched native Android code before, so really open to feedback and suggestions.

guyca commented:
@MattDavies Thanks a lot for the PR. I'm sure the community will appreciate it!

guyca commented on android/app/src/main/java/com/reactnativenavigation/activities/BottomTabActivity.java:
```diff
@@ -64,10 +64,10 @@ private void setupPages(ArrayList<Screen> screens) {
         mToolbar.setScreens(screens);
         mToolbar.setTitle(initialScreen.title);
         setSupportActionBar(mToolbar);
+        setNavigationStyle(initialScreen);
```
Out of curiosity, why did you move the call to `setNavigationStyle` to the bottom?

guyca commented on android/app/src/main/java/com/reactnativenavigation/core/objects/Screen.java:
```diff
@@ -34,6 +34,7 @@
     private static final String KEY_TOOL_BAR_STYLE = "navigatorStyle";
     private static final String KEY_STATUS_BAR_COLOR = "statusBarColor";
     private static final String KEY_TOOL_BAR_COLOR = "toolBarColor";
+    private static final String KEY_TOOL_BAR_HIDDEN = "toolBarHidden";
```
Lets change `toolBarHidden` to `navBarHidden` which is the property we [already use](https://github.com/wix/react-native-navigation#style-object-format) in iOS

MattDavies replied:
Sure, easy. There are a few other "toolBar" props we can update in a separate PR too.

MattDavies commented on android/app/src/main/java/com/reactnativenavigation/activities/BottomTabActivity.java:
```diff
@@ -64,10 +64,10 @@ private void setupPages(ArrayList<Screen> screens) {
         mToolbar.setTitle(initialScreen.title);
         setSupportActionBar(mToolbar);
+        setNavigationStyle(initialScreen);
     }
```
This had to be at the bottom to ensure the action bar was set before applying styles (else you can't show/hide).

MattDavies commented:
@guyca Changed to `navBarHidden`

guyca commented on android/app/src/main/java/com/reactnativenavigation/modal/RnnModal.java:
```diff
@@ -87,4 +105,17 @@ public void onBackPressed() {
+            mContext.runOnUiThread(new Runnable() {
+                @Override
+                public void run() {
+                    updateToolbar(mOriginalScreen, mOriginalToolbar);
```
If the Modal is dismissed, the previous toolbar already has the correct style and buttons. Do we call `updateToolbar` only for `mContext.setSupportActionBar(toolbar);` ?

MattDavies replied:
Ah. I was finding the visibility of the previous toolbar was being altered and I needed to reset it, but that might have been before I was setting the action toolbar context correctly on the modal. Let me have a play and see if I can clean this up.

guyca replied:
Great. What do you think about moving `updateToolbar` to `RnnToolbar` class? I didn't have time to check this, but I think it will be used from other places as well. 

Maybe something like this:

```
@UiThread
    public void updateToolbar(Screen screen) {
        setTitle(screen.title);
        setSupportActionBar();
        setStyle(screen);
        setupToolbarButtonsAsync(screen);
    }

    public void setSupportActionBar() {
        ((AppCompatActivity) getContext()).setSupportActionBar(this);
    }
```

Then, in `RnnModal` we would call `mToolBar.updateToolbar(mScreen);`
This also eliminates the need to keep a hard ref to `mOriginalToolbar`, a View that belongs to the previous Activity/Modal, which is a bad practice.

If we expose `setSupportActionBar` we also don't need to keep a ref to `mOriginalScreen` - seems like it simplifies the class a bit.

MattDavies commented on android/app/src/main/java/com/reactnativenavigation/activities/BottomTabActivity.java:
```diff
@@ -104,14 +104,15 @@ public boolean onCreateOptionsMenu(Menu menu) {
         super.pop(navigatorId);
         Screen screen = mScreenStacks.get(mCurrentStackPosition).pop();
-        setNavigationStyle(screen);
+        setNavigationStyle(getCurrentScreen());
```
When popping it was applying the style of the screen that had just been popped rather than the new screen.

MattDavies commented on android/app/src/main/java/com/reactnativenavigation/modal/RnnModal.java:
```diff
@@ -67,10 +75,17 @@ public void onDisplayed() {
+            ModalController.getInstance().remove();
+            super.onBackPressed();
+            return null;
+        }
```
When pop was called on the only remaining screen on a modal it didn't appear to be clearing the modal out properly.

MattDavies commented:
Hey @guyca I managed to simplify the modal logic a fair bit and fixed another couple of issues which I was hitting. I've got a couple of changes in https://github.com/MattDavies/react-native-navigation/tree/test-toolbar-hide which you might find useful for testing as well.

MattDavies commented on android/app/src/main/java/com/reactnativenavigation/modal/RnnModal.java:
```diff
@@ -66,11 +73,18 @@ public void onDisplayed() {
     }
 
     public void push(Screen screen) {
+        updateToolbar(screen, mToolBar);
```
It didn't look as though we were previously updating the toolbar style for a pushed screen within a modal.

guyca replied:
Good catch! We also miss this when pushing a screen in Activities (`BottomTabActivity` and `SingleScreenActivity`). That's why I suggested we declare `updateToolbar` in `RnnToolbar`.

guyca commented:
@MattDavies I went over #60ab611a31a5b5681e236b4f7a8ca3f0a51f6622 and seems like you're still using `toolBarHidden` instead of `navBarHidden`. Or did I miss something?

If you wish I can move `updateToolbar` to `RnnToolBar` for you. Ping me once you update the PR and I'll merge it.

PS
I'll be out of the office all of next week, So if we don't merge today I'll ask one of the other guys here to pick this up.

MattDavies commented:
Updated to move `updateToolbar` to `RnnToolbar`. It definitely reads a lot clearer now - great suggestion!

I think I lost the `navBarHidden` change somewhere along the lines...added it back.

MattDavies commented on android/app/src/main/java/com/reactnativenavigation/activities/BottomTabActivity.java:
```diff
@@ -64,10 +64,9 @@ private void setupPages(ArrayList<Screen> screens) {
-        mToolbar.setTitle(initialScreen.title);
-        setSupportActionBar(mToolbar);
+        mToolbar.updateToolbar(initialScreen);
+        setNavigationStyle(initialScreen);
```
This still sets window styles so I left it there even though there's now a slight duplication. Lets look at that in another PR?

guyca replied:
Sure, we'll address that in a separate PR. I'll merge soon. Thanks for the contribution Matt 👍

MattDavies replied:
Welcome! Great library, it's making a huge difference to some of my apps compared to the official nav. I'll keep jumping in where I can to fill gaps.
